homeassistant/components/huawei_solar/config_flow.py

```diff
@@ -0,0 +1,125 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_PORT = 502
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Required(CONF_SLAVE_IDS, default="0"): str,
+    }
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    inverter = None
+    try:
+        inverter = await AsyncHuaweiSolar.create(
+            host=data[CONF_HOST],
+            port=data[CONF_PORT],
+            slave=data[CONF_SLAVE_IDS][0],
+        )
+
+        model_name, serial_number = await inverter.get_multiple(
+            [rn.MODEL_NAME, rn.SERIAL_NUMBER]
+        )
+
+        _LOGGER.info(
+            f"Successfully connected to inverter {model_name.value} with SN {serial_number.value}"
+        )
+
+        # Also validate the other slave-ids
+        for slave_id in data[CONF_SLAVE_IDS][1:]:
+            try:
+                slave_model_name, slave_serial_number = await inverter.get_multiple(
+                    [rn.MODEL_NAME, rn.SERIAL_NUMBER], slave_id
+                )
+
+                _LOGGER.info(
+                    f"Successfully connected to slave inverter {slave_id}: {slave_model_name.value} with SN {slave_serial_number.value}"
+                )
+            except Exception as err:  # pylint: disable=broad-except
+                _LOGGER.error(f"Could not connect to slave {slave_id}")
+                raise SlaveException(f"Could not connect to slave {slave_id}") from err
+
+        # Return info that you want to store in the config entry.
+        return {"model_name": model_name.value, "serial_number": serial_number.value}
+
+    finally:
+        if inverter is not None:
+            # Cleanup this inverter object explicitly to prevent it from trying to maintain a modbus connection
+            await inverter.stop()
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Huawei Solar."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
+        errors = {}
+
+        try:
+
+            user_input[CONF_SLAVE_IDS] = list(
+                map(int, user_input[CONF_SLAVE_IDS].split(","))
+            )
+            info = await validate_input(self.hass, user_input)
+
+        except ConnectionException:
+            errors["base"] = "cannot_connect"
+        except SlaveException:
+            errors["base"] = "slave_cannot_connect"
+        except ReadException:
+            errors["base"] = "read_error"
```

Comment:
Are you sure these exceptions can happen? Since you have a broad try catch in `validate_input` that will raise a SlaveException only.

Reply:
Connectivity to these inverters is iffy, and leads to A LOT of support requests. I want to have tailored feedback to the user so that we can document each cause (and possible solutions)

Reply:
This is good, to have different exceptions. But where do you raise them? If you look at your code, you call `validate_input` where you have a broad except.

This means that you possibly only return SlaveException to this try/catch. Or do you want to catch issues with `AsyncHuaweiSolar.create(`? That is not clear, hence my comments.

https://github.com/home-assistant/core/blob/832f5f93cc03a204709b9dd5e67dea2bc1cf87c6/homeassistant/components/huawei_solar/config_flow.py#L68-L70

Reply:
Yes, the `AsyncHuaweiSolar.create()` throws these different errors.